### DISCUSSION TITLE
[e!cockpit] Can't publish

### DISCUSSION THREAD

**Discussion by Foudtz** [MichielVanwelsenaere/HomeAutomation.CoDeSys3#146] [Q&A]
Hi !

First, thanks for the work !

I've followed the CODESYS-MQTT/doc/integration.md but i can't figure out why the Publish don't work.
There is nothing in ErrorHistory, and the Publisher.done never go to TRUE value ):

Subscriber works perfectly !

Did I miss something on the Publisher's initiation ?

```
//** Handler **//
HandleMQTT(
			MQTT_IN_OUT 	:= 	MAIN_IN_OUT,
			ENABLE 			:=	xEnable,
			URL				:=	sURL,
			TIMEOUT			:= 	T#2S,
			TLS				:=	FALSE
);

IF NOT Publish_Init THEN
	Publish_Init := TRUE;
	Publisher.SetMqttInOut( MQTT_IN_OUT := MAIN_IN_OUT );
END_IF

//** Publisher - KO **//
Publisher(
	Topic				:= ADR(sTopic), 
	PayloadString 		:= ADR(sPayload),
	PublishAsString		:= TRUE,  
	Payload				:= STRING_TO_BYTE(sTopic),
	QoSIn				:= MQTT.QoS.ExactlyOnce, 
	MRetain				:= FALSE, 
	send				:= xPublish
);

IF( Publisher.done ) THEN
	xPublish := FALSE;
END_IF

//** Subscriber -  **//
IF NOT Subscriber_Init THEN
	Subscriber_Init := TRUE;
	Subscriber.SetMqttInOut( MQTT_IN_OUT := MAIN_IN_OUT );
	Collector.put( instance := Receiver );
	Receiver.initAsFindTopic( compString := ADR(sTopic), returnHit := TRUE );
END_IF

Subscriber(
	Subscribe		:=	xSubscribe,
	Topic			:=	ADR(sTopic),
	QoSSubscribe	:=	MQTT.QoS.ExactlyOnce,
	ExpectingString	:=	TRUE,
	Callback		:=	collector
);
```

Thanks for your return

**Comment by MichielVanwelsenaere** [comment]
Hi Foudtz,

you're making your own adoptation here right?
I don't see any usage of the 'xPublish' flag, are you sure it's being used correctly?
I haven't touched my publish logic for quite some time as it just works so consider my answer an educated guess :)

**Reply by Foudtz** [reply]
Hi

xPublish was manual, for testing
I've changed and used the native MQTT blocks

Thanks anyway
